packages/hebao_v1/contracts/modules/security/GuardianModule.sol

```diff
@@ -62,45 +58,61 @@ abstract contract GuardianModule is SecurityModule
         _addGuardian(wallet, guardian, group, GUARDIAN_PENDING_PERIOD);
     }
 
-    function cancelGuardianAddition(
-        address wallet,
-        address guardian
+    function addGuardiansWithTheirApproval(
+        SignedRequest.Request memory request,
+        uint    group
         )
         external
-        txAwareHashNotAllowed()
-        onlyFromWalletOrOwnerWhenUnlocked(wallet)
     {
-        controllerCache.securityStore.cancelGuardianAddition(wallet, guardian);
-        emit GuardianAdditionCancelled(wallet, guardian);
-    }
+        require(block.timestamp <= request.validUntil, "EXPIRED_SIGNED_REQUEST");
+        require(request.signers.length > 0, "EMPTY_SIGNERS");
 
-    function removeGuardian(
-        address wallet,
-        address guardian
-        )
-        external
-        txAwareHashNotAllowed()
-        onlyFromWalletOrOwnerWhenUnlocked(wallet)
-        onlyWalletGuardian(wallet, guardian)
-    {
-        _removeGuardian(wallet, guardian, GUARDIAN_PENDING_PERIOD);
+        bytes32 _txAwareHash = EIP712.hashPacked(
+            GUARDIAN_DOMAIN_SEPERATOR,
+            abi.encode(
+                ADD_GUARDIAN_TYPEHASH,
+                request.wallet,
+                address(0),
+                request.validUntil,
+                group
+            )
+        );
+
+        require(_txAwareHash == txAwareHash(), "TX_INNER_HASH_MISMATCH");
+        controllerCache.hashStore.verifyAndUpdate(request.wallet, _txAwareHash);
+
+        require(
+            _txAwareHash.verifySignatures(request.signers, request.signatures),
+            "INVALID_SIGNATURES"
+        );
+
+        address owner = Wallet(request.wallet).owner();
+        for (uint i = 0; i < request.signers.length; i++) {
+            if (request.signers[i] != owner) {
+                _addGuardian(
+                    request.wallet,
+                    request.signers[i],
+                    group,
+                    GUARDIAN_PENDING_PERIOD
+                );
+            }
+        }
     }
 
-    function addGuardianImmediately(
+    function addGuardianWA(
         SignedRequest.Request calldata request,
         address guardian,
         uint    group
         )
         external
-        onlyHaveEnoughGuardians(request.wallet)
     {
         controller().verifyRequest(
             GUARDIAN_DOMAIN_SEPERATOR,
             txAwareHash(),
             GuardianUtils.SigRequirement.OwnerAllowed,
```

Comment:
I still find it a bit strange that guardians can add/remove guardians without the owner, but I don't think there's anything wrong with it. Only that maybe guardians can sneakily add additional guardians without the owner knowing (when e.g. funds in the wallet are still low, so no reason to immediately take over the wallet), so if the owner isn't careful he can get into state where the guardians are differently than he expected in the future.

Reply:
I think it is OK. the majority of guardians can recover the wallet, so they can do whatever else.

Reply:
One improvement can be that before guardians collectively remove/add another guardian, the wallet must be locked first. How do you like the idea?